runtime/domain.c

```diff
@@ -630,28 +699,146 @@ CAMLexport void caml_reset_domain_lock(void)
   return;
 }
 
-void caml_init_domains(uintnat minor_heap_wsz) {
-  int i;
-  uintnat size;
-  void* heaps_base;
+/* minor heap initialization and resizing */
 
+static uintnat get_minor_heap_reservation_size() {
   /* sanity check configuration */
-  if (caml_mem_round_up_pages(Bsize_wsize(Minor_heap_max))
-          != Bsize_wsize(Minor_heap_max))
-    caml_fatal_error("Minor_heap_max misconfigured for this platform");
+  if (caml_mem_round_up_pages(Bsize_wsize(caml_minor_heap_max_wsz))
+          != Bsize_wsize(caml_minor_heap_max_wsz))
+    caml_fatal_error("minor_heap_max misconfigured for this platform");
+
+  return ((uintnat)Bsize_wsize(caml_minor_heap_max_wsz)
+          * Max_domains);
+}
+
+static void reserve_minor_heaps(uintnat size) {
+  void* heaps_base;
 
   /* reserve memory space for minor heaps */
-  size = (uintnat)Bsize_wsize(Minor_heap_max) * Max_domains;
   heaps_base = caml_mem_map(size, caml_sys_pagesize, 1 /* reserve_only */);
   if (heaps_base == NULL)
-    caml_fatal_error("Not enough heap memory to start up");
+    caml_fatal_error("Not enough heap memory to reserve minor heaps");
 
   caml_minor_heaps_base = (uintnat) heaps_base;
   caml_minor_heaps_end = (uintnat) heaps_base + size;
 
+  caml_gc_log("new minor heap reserved from %p to %p",
+              (value*)caml_minor_heaps_base, (value*)caml_minor_heaps_end);
+
+  for (int i = 0; i < Max_domains; i++) {
+    struct dom_internal* dom = &all_domains[i];
+
+    uintnat domain_minor_heap_area = caml_minor_heaps_base +
+      (uintnat)Bsize_wsize(caml_minor_heap_max_wsz) * (uintnat)i;
```

Comment:
It is odd that the function takes `size` as an argument but uses `caml_minor_heap_max_wsz` here. This is correct only under the assumption that `size == Bsize_wsize(caml_minor_heap_max_wsz) * Max_domains`. See my comment elsewhere about pushing `get_minor_heap_reservation_size` into this function. That way, `reserve_minor_heaps` will not take any arguments.